rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,667 @@
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ACTION_SERVER_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_take_goal_request(
+  const rcl_action_server_t * action_server,
+  rcl_action_goal_info_t * goal_info,
+  void * ros_goal);
+
+/// Send a response for a goal request to an action client using a rcl_action_server_t.
+/**
+ * This is a non-blocking call.
+ *
+ * It is the job of the caller to ensure that the type of the `ros_goal`
+ * parameter and the type associate with the action server, via the type
+ * support, match.
+ * Passing a different type produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * `goal_handle` is always owned by the calling code, but should remain constant during
+ * the call to this function.
+ *
+ * This function is thread safe so long as access to both the action server and the
+ * `goal_handle` are synchronized.
+ * That means that calling rcl_action_send_goal_response() from multiple threads is
+ * allowed, but calling rcl_action_send_goal_response() at the same time as non-thread safe
+ * action server functions is not, e.g. calling rcl_action_send_goal_response() and
+ * rcl_action_server_fini() concurrently is not allowed.
+ * Before calling rcl_action_send_goal_response() the goal handle can change and after calling
+ * rcl_action_send_goal_response() the goal handle can change, but it cannot be changed during
+ * the rcl_action_send_goal_response() call.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | Yes [1]
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] for unique pairs of action servers and responses, see above for more</i>
+ *
+ * \param[in] action_server handle to the action server that will make the goal response
+ * \param[in] goal_info a struct holding info about the goal the server is responding to
+ * \param[in] ros_goal a struct holding the goal message that the server is responding to
+ * \param[in] accepted whether or not the goal has been accepted
+ * \param[out] goal_handle a preallocated struct that is populated with goal information
+ *   if `accepted` is true, otherwise it is unmodified.
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_goal_response(
+  const rcl_action_server_t * action_server,
+  const rcl_action_goal_info_t * goal_info,
+  const void * ros_goal,
```

Comment:
Is this the original request message for the goal? Is this needed? I would have expected the `rcl_action_goal_info_t` to be enough.